#Network level hints 4: Torus network detection and split hints for torus topology

KavithaTipturMadhu opened the pull request:
This PR adds support to check if the input network topology file is a torus/mesh graph and computes the relative coordinates of the nodes in the network graph, and adds support for both minimum communicator size and minimum memory size split hints. Known limitation of the type classification algorithm is that torus links must be omitted when generating the input network topology file.  Test graph files are attached in the PR.
[meshgraphs.zip](https://github.com/pmodels/mpich/files/2477651/meshgraphs.zip)

pavanbalaji commented on src/mpi/init/netloc_util.c:
```diff
@@ -36,13 +40,14 @@ static int get_tree_attributes(netloc_topology_t topology,
         (netloc_dt_lookup_table_t *) MPL_malloc(sizeof(netloc_dt_lookup_table_t), MPL_MEM_OTHER);
     mpi_errno = netloc_get_all_host_nodes(topology, host_nodes);
 
+    MPIR_CHKPMEM_DECL(3);
```
Declarations should go to the top of the function.

KavithaTipturMadhu replied:
will fix

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -607,6 +607,93 @@ static int network_split_by_minsize(MPIR_Comm * comm_ptr, int key, int subcomm_m
+        MPIR_Process.network_attr.type == MPIR_NETLOC_NETWORK_TYPE__INVALID) {
+        color = MPI_UNDEFINED;
+    } else
+        if (MPIR_Process.network_attr.type == MPIR_NETLOC_NETWORK_TYPE__FAT_TREE ||
```
`else if` block should go on the same line.

KavithaTipturMadhu replied:
This was the whitespace checker, I think. I'll make the change and run the check one more time to confirm my suspicion.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -25,6 +26,13 @@ typedef struct {
+        struct {
+            int dimension;
+            int *geometry;
+            /*Flat index assignment in round robin manner to PUs
```
Space after `/*`

KavithaTipturMadhu replied:
Will fix

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -25,6 +26,13 @@ typedef struct {
+            int *geometry;
+            /*Flat index assignment in round robin manner to PUs
+             * that translates to a tuple, indexed by the uid of nodes */
+            int *node_coordinates;
```
Should this be an `int **` since it's for all nodes?  Also, do we want to maintain information about all nodes locally?

KavithaTipturMadhu replied:
The array is indexed by the uid of each node in the network and assigns an integer value to it, i.e., in a 2 d network with dimensions 2x2, node coordinates contains 0,1,2 & 3.

pavanbalaji replied:
Can you also answer my second question?

KavithaTipturMadhu replied:
Sorry I missed this question. The current pr does not need coordinates of all nodes to be stored in each process. I expected the entire network information to be useful in the future for rank reordering. I will update this PR to get rid of redundant information and introduce it later, if necessary.

pavanbalaji commented on src/mpi/init/netloc_util.c:
```diff
@@ -10,6 +10,20 @@
 #ifdef HAVE_NETLOC
 #include "netloc_util.h"
 #include "mpl.h"
+#include <math.h>
```
What are we relying on `math.h` for?  Is there a dependency on `libm`?

KavithaTipturMadhu replied:
I am making a call to ceilf in line 1406.

pavanbalaji replied:
Can you also answer my second question?

KavithaTipturMadhu replied:
The c++ examples have an explicit dependence on math. Otherwise, we don't seem to be linking against libm. Sorry I missed looking into this. I'll fix it.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -474,10 +479,45 @@ static int get_color_from_subset_bitmap(int node_index, int *bitmap, int bitmap_
 
-static int network_split_by_minsize(MPIR_Comm * comm_ptr, int key, int subcomm_min_size,
-                                    MPIR_Comm ** newcomm_ptr)
+static void get_prime_factors(int num, int node_dimensions, int **prime_factors, int *num_factors)
```
We have a function to get prime factors in `MPI_Dims_create` too.  Maybe we can merge them into a single function in `MPL_`?

KavithaTipturMadhu replied:
Will do.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -16,6 +16,11 @@
 
+#ifdef HAVE_NETLOC
+#include "netloc_util.h"
+#include <math.h>
```
Where's `math.h` used?

KavithaTipturMadhu replied:
This one was added by mistake, I will remove this.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -526,37 +526,124 @@ static int network_split_by_minsize(MPIR_Comm * comm_ptr, int key,
+        MPIR_Assert(num_processes_at_node != NULL);
         /* Send the count to processes */
         mpi_errno =
             MPIR_Allreduce(MPI_IN_PLACE, num_processes_at_node, num_nodes, MPI_INT,
```
Shouldn't this be an `MPID_Allreduce`?

KavithaTipturMadhu replied:
Will update this.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -756,54 +755,139 @@ static int network_split_by_min_memsize(MPIR_Comm * comm_ptr, int key, long min_
         int prev_comm_color, current_comm_color;
         int num_nodes;
         MPIR_Errflag_t errflag = MPIR_ERR_NONE;
-
```
Unnecessary white space change?

KavithaTipturMadhu replied:
Will fix.

KavithaTipturMadhu commented:
@pavanbalaji I have addressed all the review comments. There is no libm dependence now, with the ceilf operation replaced. Can you take a look at this again?

KavithaTipturMadhu commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -511,7 +511,7 @@ static int network_split_by_minsize(MPIR_Comm * comm_ptr, int key, int subcomm_m
         /* Send the count to processes */
         mpi_errno =
-            MPIR_Allreduce(MPI_IN_PLACE, num_processes_at_node, num_nodes, MPI_INT,
+            MPID_Allreduce(MPI_IN_PLACE, num_processes_at_node, num_nodes, MPI_INT,
```
Sorry about this one. Fixed.

pavanbalaji replied:
I don't see the fix.

KavithaTipturMadhu replied:
I thought you meant whitespace change. This change is from a fixup commit in the same patch list. Shall I move this to a different PR altogether?

pavanbalaji replied:
Your commit message says: `comm_split_type: split by minimum memory size`.  This patch is unrelated.

KavithaTipturMadhu replied:
Sorry this patch is from the previous PR that I have rebased on. I have modified the other patch to not have this change and moved this change to a fixup patch in this PR.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -45,5 +53,4 @@ int MPIR_Netloc_get_hostnode_index_in_tree(MPIR_Netloc_network_attributes attrib
                                            netloc_topology_t topology,
                                            netloc_node_t * network_endpoint,
                                            int *index, int *num_nodes);
-
```
Irrelevant change.

KavithaTipturMadhu replied:
Fixed.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -607,6 +607,92 @@ static int network_split_by_minsize(MPIR_Comm * comm_ptr, int key, int subcomm_m
 }
+
+static int network_split_by_min_memsize(MPIR_Comm * comm_ptr, int key, long min_mem_size,
+                                        MPIR_Comm ** newcomm_ptr)
```
Why isn't `min_memsize` basically just `minsize`, but by calculating the number of nodes based on the available memory?

pavanbalaji replied:
I think it's reasonable to assume that.  I don't know of any reasonable system where that's not true, and I'm sure we'll run into a number of other weird problems if that weren't true.

pavanbalaji commented on test/mpi/comm/cmsplit_type.c:
```diff
@@ -22,7 +22,7 @@ static const char *split_topo[] = {
 
 static const char *split_topo_network[] = {
-    "switch_level:2", "subcomm_min_size:2", NULL
+    "switch_level:2", "subcomm_min_size:2", "min_mem_size:512", NULL
```
What's the unit here?

KavithaTipturMadhu replied:
The unit of memory is in bytes. I'll add comments and description in the PR stating this.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -860,6 +1010,10 @@ int MPIR_Comm_split_type_network_topo(MPIR_Comm * comm_ptr, int key, const char
+    } else if (!strncmp(hintval, ("torus_size:"), strlen("torus_size:"))
+               && *(hintval + strlen("torus_size:")) != '\0') {
+        int min_size = atol(hintval + strlen("torus_size:"));
+        mpi_errno = network_split_by_min_torus_size(comm_ptr, key, min_size, newcomm_ptr);
```
How is `torus_size` different from `subcomm_min_size`?

KavithaTipturMadhu replied:
Torus size is based on the network properties, and assumes uniform distribution of processes among all nodes of the network.  ```subcomm_min_size``` makes no assumptions about the number of MPI ranks/node in the network.

pavanbalaji replied:
Why would the user care about whether it's a torus or a fat tree or some other topology?  Wouldn't they want just a collection of processes that are "close to each other"?  In that sense, `subcomm_min_size` is all we need: that needs to be applied to fat-tree or torus or other topologies internally without exposing that information to the user.

KavithaTipturMadhu replied:
That makes sense, I'll get rid of this. We decided to go with switch-level hint for fat trees, which is also covered by subcomm_min_size, do you think I should get rid of that as well?

pavanbalaji replied:
No, I think the `switch-level` hint is more like "give me all the processes on the first level of the switch".  It doesn't specify how many processes the application needs.  An equivalent for the torus would be something like "give me all the processes that share the x-dimension of the torus" (which we can provide with the current topology information that we already query), or "give me all the processes that are on the same rack" (which might be harder to query with the current hwloc/netloc).

KavithaTipturMadhu replied:
Torus size is not exactly in the same vein, but it also does not specify how many processes the application needs, but covers as many nodes as the user requests, while maximizing the covered dimensions. But I get your point about how this might be very topology aware and obscure to the user. I will remove the torus size hint and add torus dimension hint instead.

KavithaTipturMadhu replied:
I have updated the patches for torus dimension instead of torus size.

pavanbalaji commented:
Can you also move all the fixup comments to the start?

KavithaTipturMadhu commented:
> Can you also move all the fixup comments to the start?

I have committed the patches in that order, with fixups coming in first. I have no idea why the web interface shows patches in a different order. I'll push this one more time and check.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -25,6 +26,13 @@ typedef struct {
+            int *geometry;
+            /* Flat index of the node the current
+             * rank is mapped to */
+            int current_node_coordinates;
```
Can you use the name `node_idx` instead of `current_node_coordinates`, to make this less confusing?

KavithaTipturMadhu replied:
Done

KavithaTipturMadhu commented:
I have addressed all comments from you, with the minmemsize hint refactored to use split_mem_size. This gets rid of the torus minmemsize support patch, and the refactoring merged in the patch from the PR rebased on for minmemsize support. Can you take a look at the patches again?

KavithaTipturMadhu commented:
I have updated the commit messages and added sign-offs as suggested. Can you look at them once?

pavanbalaji commented:
Please run the tests on more time before merging.

KavithaTipturMadhu commented:
test:jenkins/ch3/all
test:jenkins/ch4/all

KavithaTipturMadhu commented:
@pavanbalaji I spotted an unused variable in https://github.com/pmodels/mpich/pull/3204/commits/27635d56d27bf5c7d6462e1cef7f4b85486c6ff9 which I removed. There was also a bug which was fixed in line 635, which adds a null check for ```node_comm```. ```node_comm``` is NULL when the local size of the communicator is 1, and I did not have a check for this earlier, resulting in a seg fault when the test is run with 1 process. I have removed your signoff from the commit. Can you take a look at this?

KavithaTipturMadhu commented:
Starting tests again while I await review, since there were some minor changes.
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
[Fortran ofi tests have been reported to be failing](https://github.com/pmodels/mpich/issues/3421). ```gather_big``` test is timing out on ofi, and is unrelated to the PR. However, I will check why the test is failing and go ahead and merge this PR.
